### PR TITLE
feat: Download files progressively only when needed 

### PR DESCRIPTION
**Pull request by doubleface**
Now the launcher will only have to call saveFiles directly for all the
given files,  which will avoid to download all the files before calling
saveFiles. (PR in flagship app to follow)

The downloadAndFormatFile option callback has been added to allow
saveFiles to download the corresponding file in the worker when needed

For optimization, an index of already downloaded files is created on
each call to saveFiles. Then, when saveFiles is called with multiple
entries, the need to download a file will be detected quickly.

Also clean some code and added jsdoc as much as possible now.

### CODE REVIEW

File: packages/cozy-clisk/src/launcher/saveFiles.js [hunk 1]
```diff
@@ -154,29 +164,17 @@ const saveEntry = async function (client, entry, options) {
...
-    options,
-    'sourceAccountOptions.sourceAccountIdentifier'
-  )
-  if (!sourceAccountIdentifier) {
-    log.warn(
-      'No sourceAccountIdentifier is defined in options, file deduplication will be based on file path'
-    )
+    throw new Error('Savefiles : No slug is defined for the current connector')
   }
 }
```
Comment (Crash--) [inline packages/cozy-clisk/src/launcher/saveFiles.js:250]: should be tagged as a BREAKING CHANGE even if you think that all clisk will not be impacted by this change.

File: packages/cozy-clisk/src/launcher/saveFiles.js [hunk 2]
```diff
@@ -504,6 +464,15 @@ function attachFileToEntry(entry, fileDocument) {
...
 
+/**
+ * Get the full path of a given file, from it's folder path and file or entry
+ *
+ * @param {Object} arg - arg option object
+ * @param {import('cozy-client/types/types').FileDocument} [arg.file] - io.cozy.files object
+ * @param {saveFilesEntry} [arg.entry] - saveFiles entry
+ * @param {saveFileOptions} arg.options - io.cozy.files object
+ * @returns {string | undefined} - file full path
```
Comment (Crash--) [inline packages/cozy-clisk/src/launcher/saveFiles.js:474]: this is not part of this PR, but its weird to return undefined from this kind of method.

File: packages/cozy-clisk/src/launcher/saveFiles.js [hunk 3]
```diff
@@ -1,23 +1,89 @@
+// @ts-check
 import Minilog from '@cozy/minilog'
 import get from 'lodash/get'
+import set from 'lodash/set'
```
Comment (Crash--) [inline packages/cozy-clisk/src/launcher/saveFiles.js:4]: Can't we avoid the use of lodash set? 

Remember: We should go away from lodash. It'll reduce the size of the content script ;)

**Review by Crash-- — APPROVED** [review]
Once commits are squashed / removed, let's merge :)

File: packages/cozy-clisk/src/launcher/saveFiles.js (hunk 3 above)
Comment (doubleface) [reply packages/cozy-clisk/src/launcher/saveFiles.js:4]: It is allways possible but a lot less elegant and readable I think. But let's do this.

File: packages/cozy-clisk/src/launcher/saveFiles.js (hunk 1 above)
Comment (doubleface) [reply packages/cozy-clisk/src/launcher/saveFiles.js:256]: Done

File: packages/cozy-clisk/src/launcher/saveFiles.js (hunk 3 above)
Comment (doubleface) [reply packages/cozy-clisk/src/launcher/saveFiles.js]: Done

File: packages/cozy-clisk/src/launcher/saveFiles.js [hunk 4]
```diff
@@ -104,14 +104,6 @@ const saveFiles = async (client, entries, folderPath, options) => {
     existingFilesIndex: options.existingFilesIndex
   }
```
Comment (Crash--) [inline packages/cozy-clisk/src/launcher/saveFiles.js:106]: this commit should be tagged as a breaking change

File: packages/cozy-clisk/src/launcher/saveFiles.js [hunk 5]
```diff
@@ -7,17 +7,27 @@ import retry from 'bluebird-retry'
 const log = Minilog('saveFiles')
```
Comment (Crash--) [inline packages/cozy-clisk/src/launcher/saveFiles.js:8]: This is not the right semantic see https://github.com/cozy/cozy-guidelines/#breaking-change 🙏

Comment (doubleface) [reply packages/cozy-clisk/src/launcher/saveFiles.js:8]: Now it should be better

Comment (Crash--) [reply packages/cozy-clisk/src/launcher/saveFiles.js:8]: Nop. No space before the `:`

Comment (doubleface) [reply packages/cozy-clisk/src/launcher/saveFiles.js:8]: This time is the right one
